jax/_src/numpy/lax_numpy.py

```diff
@@ -3583,28 +3564,39 @@ def array(object, dtype=None, copy=True, order="K", ndmin=0):
   # check if the given dtype is compatible with JAX
   lax._check_user_dtype_supported(dtype, "array")
 
+  # Here we make a judgment call: we only return a weakly-typed array when the
+  # input object itself is weakly typed. That ensures asarray(x) is a no-op whenever
+  # x is weak, but avoids introducing weak types with something like array([1, 2, 3])
```

Comment:
Would it be feasible/desirable for `asarray(x)` always to return a strong type, even if its input is weakly typed? That is, not to ensure that `asarray(x)` is a no-op when `x` is weak? I ask in the spirit of having the simplest policy possible.

Reply:
I'm exploring that as well, but didn't include that in this PR. The behavior described in the comment is the current behavior (the comment is moved from further down)

Reply:
We could go two ways here: either have `jnp.array` always return strong types, or have `jnp.array` always respect the weak type coming from the type promotion of the inputs. I'm still unsure which would be best, and I want to run the full internal test suite on both versions to help think through the consequences.